crates/lib/src/qpu/experimental/randomized_measurements.rs

```diff
@@ -0,0 +1,628 @@
+                qubit_randomization.measurement.clone(),
+            ));
+        }
+        Ok(program)
+    }
+
+    /// Given a map of qubits to seed values, construct the parameters required
+    /// to randomize measurements accordingly.
+    pub fn to_parameters(&self, seed_values: &HashMap<Qubit, PrngSeedValue>) -> Result<Parameters> {
+        let mut parameters = HashMap::new();
+        parameters.insert(
+            RANDOMIZED_MEASUREMENT_SOURCE.to_string().into_boxed_str(),
+            self.unitary_set.to_parameters()?,
+        );
+
+        for qubit_randomization in &self.qubit_randomizations {
+            let seed_value = seed_values
+                .get(&qubit_randomization.measurement.qubit)
+                .ok_or_else(|| Error::MissingSeed(qubit_randomization.measurement.qubit.clone()))?;
+            parameters.insert(
+                qubit_randomization
+                    .seed_declaration
+                    .name
+                    .clone()
+                    .into_boxed_str(),
+                vec![seed_value.as_f64],
+            );
+            parameters.insert(
+                qubit_randomization
+                    .destination_declaration
+                    .name
+                    .clone()
+                    .into_boxed_str(),
+                std::iter::repeat(0.0)
+                    .take(self.unitary_set.parameters_per_unitary())
+                    .collect(),
+            );
+        }
+
+        Ok(parameters)
+    }
+
+    /// Given a map of qubits to seed values, return the random indices that
+    /// were played on each qubit during program execution.
+    #[must_use]
+    pub fn get_random_indices(
+        &self,
+        seed_values: &HashMap<Qubit, PrngSeedValue>,
+        shot_count: u32,
+    ) -> HashMap<Qubit, Vec<u8>> {
+        seed_values
+            .iter()
+            .map(|(qubit, seed_value)| {
+                (
+                    qubit.clone(),
+                    super::random::choose_random_real_sub_region_indices(
+                        *seed_value,
+                        0,
+                        shot_count,
+                        self.unitary_count_as_u8,
+                    ),
+                )
+            })
+            .collect()
+    }
+}
+
+/// A set of unitaries, each of which may be expressed as a set of Quil
+/// instructions.
+#[derive(Debug, Clone)]
+pub enum UnitarySet {
+    /// A set of unitaries expressed as a sequence of the following gate
+    /// operations:
+    ///
+    /// RZ(angle_0)-RX(pi/2)-RZ(angle_1)-RX(pi/2)-RZ(angle_2).
+    ///
+    /// The unitaries are stored in a 2D array where each row represents
+    /// a single unitary expressed as the three RZ angles.
+    Zxzxz(Array2<f64>),
+}
+
+impl UnitarySet {
+    /// Attempt to create a new instance of [`UnitarySet`] from a 2D array
+    /// of unitaries. The array must have three columns.
+    pub fn try_new_zxzxz(unitaries: Array2<f64>) -> Result<UnitarySet> {
```

Comment:
One way to avoid it would be to make `UnitarySet` a private type (and rename to something like `UnitarySetInner`), and make a `pub struct UnitarySet(UnitarySetInner);` as an opaque wrapper around it.